src/cpufreq.c

```diff
@@ -26,39 +44,213 @@
 
 #define MODULE_NAME "cpufreq"
 
-static int num_cpu = 0;
+// Configuration options
+static const char *config_keys[] =
+{
+	"ReportDistribution",
+	"ReportByCpu"
+};
+static int config_keys_num = STATIC_ARRAY_SIZE (config_keys);
+
+// Global variables
+
+static _Bool report_distribution = 0;  // option
+static _Bool report_by_cpu = 0;        // option
+
+static size_t num_cpu = 0;  // number of cpus
+
+// Variables used when CPU frequency distributions are calculated
+// that summarize the use for all CPUs together.
+static long int *hertz_all_cpus = NULL;  // list of available frequencies
+static derive_t *time_all_cpus = NULL;   // time spent at each frequency
+static size_t hertz_size = 0;
+static size_t reported_last_run = 0;
+
+// Functions
+
+static int cpufreq_config (const char *key, const char *value)
+{
+	if (strcasecmp (key, "ReportDistribution") == 0)
+		report_distribution = IS_TRUE(value);
+	else if (strcasecmp (key, "ReportByCpu") == 0)
+		report_by_cpu = IS_TRUE(value);
+	else return (-1);
+
+	return (0);
+}
+
 
 static int cpufreq_init (void)
 {
-        int status;
-	char filename[256];
+	char filename[128];
+	int status;
+	_Bool done;
 
-	num_cpu = 0;
+	// Determine number of cpus. Here, just check whether cpus are
+	// there. since cpus can be hot-plugged, cpufreq could be absent
+	// at this particualr time moment
 
-	while (1)
+	done = 0;
+	for (num_cpu = 0; !done; ++num_cpu)
 	{
 		status = ssnprintf (filename, sizeof (filename),
-				"/sys/devices/system/cpu/cpu%d/cpufreq/"
-				"scaling_cur_freq", num_cpu);
+							"/sys/devices/system/cpu/cpu%zu",
```

Comment:
Please keep the existing indentation intact. This only lines out if you use 4 space tabs. It also makes it hard to see what this actual change does.